infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/RuntimeMachines.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker;
+
+import static java.util.Collections.emptyMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.commons.annotation.Nullable;
+
+/**
+ * Holds machines during runtime lifecycle.
+ *
+ * @author Alexander Garagatyi
+ */
+public class RuntimeMachines {
+
+  private Map<String, DockerMachine> machines;
+
+  public RuntimeMachines() {
+    this.machines = new HashMap<>();
+  }
+
+  /** Returns machines map. If runtime start is canceled returns empty map. */
+  public synchronized Map<String, DockerMachine> getMachines() {
+    return machines != null ? machines : emptyMap();
```

Comment:
As I understand here it is better to return the copy of original machines, because of guarantee that it won't be modified by `RuntimeMachines` and you can iterate on it without risking of concurrent modification.

Reply:
The copy is done [here](https://github.com/garagatyi/che/blob/647f982342945f2fa83cf41c03db7d4f04a75697/infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java#L268-L268), whereas this component is kinda internal, so I would leave it as is, for now